pkg/adsc/delta.go

```diff
@@ -0,0 +1,739 @@
+		// No need to ack and type check for debug types
+		return nil
+	}
+	for _, r := range d.Resources {
+		if d.TypeUrl != r.Resource.TypeUrl {
+			deltaLog.Errorf("Invalid response: mismatch of type url: %v vs %v", d.TypeUrl, r.Resource.TypeUrl)
+			continue
+		}
+		err := c.trigger(ctx, d.TypeUrl, r, EventAdd)
+		if err != nil {
+			return err
+		}
+		parentKey := resourceKey{
+			Name:    r.Name,
+			TypeURL: r.Resource.TypeUrl,
+		}
+		c.establishResource(parentKey)
+		if ctx.nack != nil {
+			rejects = append(rejects, ctx.nack)
+			// On NACK, do not apply resource changes
+			continue
+		}
+
+		remove, add := c.tree[parentKey].Children.Diff(ctx.sub)
+		for _, key := range add {
+			if _, f := allAdds[key.TypeURL]; !f {
+				allAdds[key.TypeURL] = set.New[string]()
+			}
+			allAdds[key.TypeURL].Insert(key.Name)
+			c.relate(parentKey, key)
+		}
+		for _, key := range remove {
+			if _, f := allRemoves[key.TypeURL]; !f {
+				allRemoves[key.TypeURL] = set.New[string]()
+			}
+			allRemoves[key.TypeURL].Insert(key.Name)
+			c.unrelate(parentKey, key)
+		}
+	}
+	for _, r := range d.RemovedResources {
+		key := resourceKey{
+			Name:    r,
+			TypeURL: d.TypeUrl,
+		}
+		removed := &discovery.Resource{
+			Name: r,
+		}
+		err := c.trigger(ctx, d.TypeUrl, removed, EventDelete)
+		if err != nil {
+			return err
+		}
+
+		if _, f := allRemoves[key.TypeURL]; !f {
+			allRemoves[key.TypeURL] = set.New[string]()
+		}
+		allRemoves[key.TypeURL].Insert(key.Name)
+		c.drop(key)
+	}
+	c.send(resourceKey{TypeURL: d.TypeUrl}, d.Nonce, joinError(rejects))
+	for t, sub := range allAdds {
+		unsub, f := allRemoves[t]
+		if f {
+			delete(allRemoves, t)
+		}
+		c.update(t, sub, unsub, d)
+	}
+	return nil
+}
+
+func joinError(rejects []error) error {
+	var e []string
+	for _, r := range rejects {
+		if r == nil {
+			continue
+		}
+		e = append(e, r.Error())
+	}
+	if len(e) == 0 {
+		return nil
+	}
+	return errors.New(strings.Join(e, "; "))
+}
+
+// establishResource sets up the relationship for a resource we received.
+func (c *Client) establishResource(key resourceKey) {
+	// Check if we have a watch for this resource
+	parentNode, f := c.tree[key]
+	if !f {
+		parentNode = resourceNode{
+			Parents:  make(keySet),
+			Children: make(keySet),
+		}
+		c.tree[key] = parentNode
+	}
+
+	// Check if we have a Watch for all "*" resources, and if so, this specific resource is a child
+	// of that watch.
+	wildcardKey := resourceKey{TypeURL: key.TypeURL}
+	wildNode, wildFound := c.tree[wildcardKey]
+	if wildFound {
+		wildNode.Children.Insert(key)
+		parentNode.Parents.Insert(wildcardKey)
+	}
+
+	if !f && !wildFound {
+		// We are receiving an unwanted resource, silently ignore it.
+		deltaLog.Debugf("Received unsubscribed resource: %v, %v", key, c.tree)
+	}
+}
+
+func (c *Client) relate(parent, child resourceKey) {
+	parentNode, f := c.tree[parent]
+	if !f {
+		deltaLog.Fatalf("Failed to relate resource: unknown parent: %v, %v", parent, c.tree)
```

Comment:
mark as error?

Reply:
I think this actually should not happen; perhaps error may not be enough?